### PR TITLE
Adding gaussian quadrature integrator

### PR DESCRIPTION
**Pull request by ConnorStoneAstro**
A gaussian quadrature integrator allows for pixel brightnesses to be more accurately determined with fewer evaluations than simple supersampling. See further information on Gaussian quadrature here: https://en.wikipedia.org/wiki/Gaussian_quadrature

### CODE REVIEW

File: src/caustics/utils.py [hunk 1]
```diff
@@ -115,6 +116,97 @@ def get_meshgrid(
...
+
+    Args:
+        n (int): The number of quadrature points in each dimension.
+        p (Tensor): The pixelscale.
+        dtype (torch.dtype): The desired data type of the tensor.
+        device (torch.device): The device on which to create the tensor.
+
+    Returns:
+        Tuple[Tensor, Tensor, Tensor]: The generated meshgrid as a tuple of Tensors.
+    """
```
Comment (lsetiawan) [inline src/caustics/utils.py:132]: Just noticed that this docstring style is not numpydoc. Please start following this guide https://numpydoc.readthedocs.io/en/latest/format.html so that we're all consistent 😄 Thanks!

Comment (ConnorStoneAstro) [reply src/caustics/utils.py:132]: Good catch! Old habits die hard

File: src/caustics/utils.py [hunk 2]
```diff
@@ -115,6 +116,124 @@ def get_meshgrid(
...
+    w = torch.tensor(weights, dtype=dtype, device=device)
+    a = p * torch.tensor(abscissa, dtype=dtype, device=device) / 2.0
+    X, Y = torch.meshgrid(a, a, indexing="xy")
+
+    W = torch.outer(w, w) / 4.0
+
+    X, Y = X.reshape(-1), Y.reshape(-1) # flatten
+    return X, Y, W.reshape(-1)
+
+def get_pixel_quad_integrator_grid(
```
Comment (AlexandreAdam) [inline src/caustics/utils.py:151]: We should not name function `get_something`, instead we should just call the function `something`. This helps readability. Why not integrate this function into the `pixel_quad_integrator`?

**Review by AlexandreAdam — CHANGES_REQUESTED** [review]
This is a super cool addition! The changes I request are mainly about the names of the functions and moving the implementation of the interpolation scheme to the source brightness method rather than at the level of the simulator.

File: src/caustics/sims/lens_source.py [hunk 3]
```diff
@@ -161,18 +157,52 @@ def forward(
...
+                else:
+                    mu = self.source.brightness(*self.grid, params)
         else:
             # Source is not added to the scene
             mu = torch.zeros_like(self.grid[0])
 
         # Sample the lens light
         if lens_light and self.lens_light is not None:
-            mu += self.lens_light.brightness(*self.grid, params)
+            if quad_level is not None and quad_level > 1:
```
[the remark this answers is not in the document]
Comment (ConnorStoneAstro) [reply src/caustics/sims/lens_source.py:194]: I disagree about it being part of the source brightness method. The brightness method is fundamentally asking the question "what is the brightness at these x,y coordinates?", however the simulator is asking the question "what is the brightness in these pixels?". If we put the integrator in the brightness function then we are imposing the idea of pixels on the user at too early of a stage.

File: src/caustics/utils.py [hunk 4]
```diff
@@ -115,6 +116,124 @@ def get_meshgrid(
...
+    abscissaX, abscissaY, weight = _quad_table(
+        quad_level, pixelscale, dtype, device
+    )
+
+    # Gaussian quadrature evaluation points
+    Xs = torch.repeat_interleave(X[..., None], quad_level ** 2, -1) + abscissaX
+    Ys = torch.repeat_interleave(Y[..., None], quad_level ** 2, -1) + abscissaY
+    return Xs, Ys, weight
+
+def pixel_quad_integrator(
```
[the remark this answers is not in the document]
Comment (AlexandreAdam) [reply src/caustics/utils.py:202]: Oh I see, so you see the Gaussian quadrature integrator as a replacement of something like average pooling? Currently we do not have any model that work on irregular pixel grids, and so I am wondering what this brings. Is it a better integrator for sharp sources like the Sersic model?

In that case my second comment is wrong as you pointed out, this definitely belongs in the simulator at the same level as average pooling.

[the remark this answers is not in the document]
Comment (ConnorStoneAstro) [reply src/caustics/utils.py:202]: Yes, it is basically average pooling that converges faster. Average pooling is basically integrating with the midpoint rule. One step better is simpsons rule, and two steps better is gaussian quadrature :)

[the remark this answers is not in the document]
Comment (AlexandreAdam) [reply src/caustics/utils.py]: Wait I missed something, we have an upsampling layer in there? Ideally, we would be evaluating the source model at the same resolution as the PSF so that only downsampling is performed in the forward model. Unless the user specifies a specific pixel size for a pixelated source model, but then the situation is a bit more complicated. 

And I am all for the gaussian_quadrature setup, as long as the default is average_pooling. I care about this in a somewhat pedantic fashion because of the linearity of the simulator with respect to the source parameters (for pixelated models).

[the remark this answers is not in the document]
Comment (AlexandreAdam) [reply src/caustics/utils.py]: Ok let's separate two things. For a pixelated model the situation is a bit complicated. But for Sersic, you basically use the Gaussian quadrature to pixelate the model at the PSF resolution is that correct? But then I feel like we should have average pooling to go back to the observation resolution

[the remark this answers is not in the document]
Comment (AlexandreAdam) [reply src/caustics/utils.py]: Ok I went back to read the full code and this is what is going on: pixelate at PSF resolution and then convolve and then pixelate at observation resolution. Sorry my bad I was a bit slow here. 

But then my question about the pixelated source remains. How do you handle this case?

[the remark this answers is not in the document]
Comment (ConnorStoneAstro) [reply src/caustics/utils.py]: Right, the upsampling should only be used if the PSF is upsampled, otherwise it is better to use quadrature for the accuracy/evaluation ratio. The way the code stands right now, the user can choose very freely between the two. Here are the valid configurations:

- no PSF convolution, user can freely choose upsample factor, or quadrature factor, or both at will 
- with PSF convolution, user should set upsample factor to the upsampling of the PSF, quadrature can be set to any level

So the quadrature can always be set to anything and it is valid. The upsampling/avgpooling combo has to match the PSF if there is a PSF. 

As for linearity, the gaussian quadrature is also linear. It is actually equivalent to a convolution with stride equal to the kernel size, which is the same as kernelized block pooling which is what I do.

[the remark this answers is not in the document]
Comment (ConnorStoneAstro) [reply src/caustics/utils.py]: A pixelated source is treated like any other source. Since a pixelated source is defined by bilinear interpolation, it can also be evaluated an integrated at any set of coordinates. That said, because a pixelated source is linear at the pixel scale, there is no reason to do sub-pixel integration as the midpoint rule is exact for linear integrals.

[the remark this answers is not in the document]
Comment (AlexandreAdam) [reply src/caustics/utils.py]: Ok make sense. Thanks for the explanation!

[the remark this answers is not in the document]
Comment (ConnorStoneAstro) [reply src/caustics/utils.py]: This is some good discussion. I'll add a rewritten version to the documentation :)

**Review by AlexandreAdam — APPROVED** [review]
